selection.go

```diff
@@ -0,0 +1,267 @@
+		&stun.MessageIntegrity{
+			Key: []byte(a.remotePwd),
+		},
+		&stun.Fingerprint{},
+	)
+
+	if err != nil {
+		a.log.Error(err.Error())
+		return
+	}
+
+	a.log.Tracef("ping STUN from %s to %s\n", local.String(), remote.String())
+	if overflow := len(a.pendingBindingRequests) - maxPendingBindingRequests; overflow > 1 {
+		a.log.Debugf("Discarded %d pending binding requests, pendingBindingRequests is full", overflow)
+		a.pendingBindingRequests = a.pendingBindingRequests[overflow:]
+	}
+
+	a.pendingBindingRequests = append(a.pendingBindingRequests, bindingRequest{
+		transactionID: transactionID,
+		destination:   remote.addr(),
+	})
+	a.sendSTUN(msg, local, remote)
+}
+
+func (s *ControlledSelector) HandleSucessResponse(m *stun.Message, local, remote *Candidate, remoteAddr net.Addr) {
+	// TODO according to the standard we should specifically answer a failed nomination:
+	// https://tools.ietf.org/html/rfc8445#section-7.3.1.5
+	// If the controlled agent does not accept the request from the
+	// controlling agent, the controlled agent MUST reject the nomination
+	// request with an appropriate error code response (e.g., 400)
+	// [RFC5389].
+
+	ok, pendingRequest := s.agent.handleInboundBindingSuccess(m.TransactionID)
+	if !ok {
+		s.agent.log.Errorf("discard message from (%s), invalid TransactionID %s", remote, m.TransactionID)
+		return
+	}
+
+	transactionAddr := pendingRequest.destination
+
+	// Assert that NAT is not symmetric
+	// https://tools.ietf.org/html/rfc8445#section-7.2.5.2.1
+	if !addrEqual(transactionAddr, remoteAddr) {
+		s.agent.log.Debugf("discard message: transaction source and destination does not match expected(%s), actual(%s)", transactionAddr, remote)
+		return
+	}
+
+	s.agent.log.Tracef("inbound STUN (SuccessResponse) from %s to %s", remote.String(), local.String())
+	s.agent.addValidPair(local, remote)
+
+	if bytes.Equal(s.validateNominationTxID, m.TransactionID) {
+		s.agent.sendBindingSuccess(s.nominationTxID, local, remote)
+	}
+}
+
+func (s *ControlledSelector) HandleBindingRequest(m *stun.Message, local, remote *Candidate) {
+	_, useCandidate := m.GetOneAttribute(stun.AttrUseCandidate)
+
+	if useCandidate {
+		// https://tools.ietf.org/html/rfc8445#section-7.3.1.5
+		p := s.agent.findValidPair(local, remote)
+
+		if p != nil {
+			// If the state of this pair is Succeeded, it means that the check
+			// previously sent by this pair produced a successful response and
+			// generated a valid pair (Section 7.2.5.3.2).  The agent sets the
+			// nominated flag value of the valid pair to true.
+			s.agent.setSelectedPair(p)
+			s.agent.sendBindingSuccess(m.TransactionID, local, remote)
+		} else {
+			p = newCandidatePair(local, remote, false)
+			s.pingNominatedCandidate(p, m.TransactionID)
+		}
+	} else {
+		s.agent.sendBindingSuccess(m.TransactionID, local, remote)
+	}
+}
+
+func (s *ControlledSelector) pingNominatedCandidate(pair *candidatePair, nominationTxID []byte) {
+	a := s.agent
```

Comment:
(nit please push back if you disagree) I wouldn't put variables on the stack if you don't have to. IMO it makes things harder to grok (more LoC and just more state I need to keep track of when reading)

Reply:
no problem! done